paychmgr/simple.go

```diff
@@ -41,106 +250,268 @@ func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, am
+	}
+
+	go ca.waitForAddFundsMsg(from, to, mcid, cb)
+
+	return &mcid, nil
+}
+
+// waitForAddFundsMsg waits for mcid to appear on chain and returns error, if any
+func (ca *channelAccessor) waitForAddFundsMsg(from address.Address, to address.Address, mcid cid.Cid, cb onCompleteFn) {
+	err := ca.waitAddFundsMsg(from, to, mcid)
+	ca.msgWaitComplete(mcid, err, cb)
 }
 
-// WaitForAddFundsMsg waits for mcid to appear on chain and returns error, if any
-// TODO: wait outside the store lock!
-//  (tricky because we need to setup channel tracking before we know it's address)
-func (pm *Manager) waitForAddFundsMsg(ctx context.Context, mcid cid.Cid) {
-	defer pm.store.lk.Unlock()
-	mwait, err := pm.state.StateWaitMsg(ctx, mcid, build.MessageConfidence)
+func (ca *channelAccessor) waitAddFundsMsg(from address.Address, to address.Address, mcid cid.Cid) error {
+	mwait, err := ca.api.StateWaitMsg(ca.waitCtx, mcid, build.MessageConfidence)
 	if err != nil {
 		log.Error(err)
+		return err
 	}
 
 	if mwait.Receipt.ExitCode != 0 {
-		log.Errorf("voucher channel creation failed: adding funds (exit code %d)", mwait.Receipt.ExitCode)
+		err := xerrors.Errorf("voucher channel creation failed: adding funds (exit code %d)", mwait.Receipt.ExitCode)
+		log.Error(err)
+
+		ca.lk.Lock()
+		defer ca.lk.Unlock()
+
+		ca.mutateChannelInfo(from, to, func(channelInfo *ChannelInfo) {
+			channelInfo.PendingAmount = big.NewInt(0)
+			channelInfo.AddFundsMsg = nil
+		})
+
+		return err
 	}
-}
 
-func (pm *Manager) GetPaych(ctx context.Context, from, to address.Address, ensureFree types.BigInt) (address.Address, cid.Cid, error) {
-	pm.store.lk.Lock() // unlock only on err; wait funcs will defer unlock
-	var mcid cid.Cid
-	ch, err := pm.store.findChan(func(ci *ChannelInfo) bool {
-		if ci.Direction != DirOutbound {
-			return false
-		}
-		return ci.Control == from && ci.Target == to
+	ca.lk.Lock()
+	defer ca.lk.Unlock()
+
+	// Store updated amount
+	ca.mutateChannelInfo(from, to, func(channelInfo *ChannelInfo) {
+		channelInfo.Amount = types.BigAdd(channelInfo.Amount, channelInfo.PendingAmount)
+		channelInfo.PendingAmount = big.NewInt(0)
+		channelInfo.AddFundsMsg = nil
 	})
+
+	return nil
+}
+
+// Change the state of the channel in the store
+func (ca *channelAccessor) mutateChannelInfo(from address.Address, to address.Address, mutate func(*ChannelInfo)) {
+	channelInfo, err := ca.store.OutboundActiveByFromTo(from, to)
+
+	// If there's an error reading or writing to the store just log an error.
+	// For now we're assuming it's unlikely to happen in practice.
+	// Later we may want to implement a transactional approach, whereby
+	// we record to the store that we're going to send a message, send
+	// the message, and then record that the message was sent.
 	if err != nil {
-		pm.store.lk.Unlock()
-		return address.Undef, cid.Undef, xerrors.Errorf("findChan: %w", err)
+		log.Errorf("Error reading channel info from store: %s", err)
```

Comment:
will OutboundActiveByFromTo always return a channel info, even if it errors? If not, we need to return here, or something

Reply:
So I think actually the correct thing to do here is to panic - because if the datastore is not working the system is in an undefined state

Reply:
https://github.com/filecoin-project/lotus/pull/2870

Reply:
Magik pointed out that panicking is worse than just bailing: https://github.com/filecoin-project/lotus/pull/2871